File: force-field/ruby/Particle.rb
require './Vector.rb'

class Particle
  attr_accessor :position
  attr_accessor :mass
  attr_accessor :velocity
  
  def initialize(pos, mass)
    @position = pos
    @mass = mass
    @velocity = Vector.new 0, 0
  end
  
  def forceFrom(p)
    distance = p.position.subtract @position
    magnitude = @mass * p.mass / distance.dot(distance)
    return distance.normalize().scale magnitude
  end
  
  def applyForce(force, timeDelta)
    velocity.addInPlace force.scale (timeDelta / @mass.abs)
    position.addInPlace velocity.scale timeDelta
  end
  
  def to_s
    "{Particle mass: #{@mass} velocity: #{@velocity} position: #{@position}}"
  end
end

File: force-field/rust/particle.rs
use std::fmt::{Display, Formatter, Result};

use super::vector::Vector;

pub struct Particle {
  pub position: Vector,
  pub mass: f64,
  velocity: Vector,
}

impl Particle {
  pub fn new(position: Vector, mass: f64) -> Particle {
    Particle {
      position: position,
      mass: mass,
      velocity: Vector::new(),
    }
  }

  pub fn force_from(&self, other: &Particle) -> Vector {
    let distance = self.position - other.position;

    // G*m1*m1/d^2, assuming G=1
    let magnitude = self.mass * other.mass / distance.dot(&distance);

    let mut vector = distance.normalize();
    vector.scale(magnitude);

    vector
  }

  pub fn apply_force(&mut self, force: Vector, time_delta: f64) {
    self.velocity += (time_delta / self.mass.abs()) * force;
    self.position += time_delta * self.velocity;
  }
}

impl Display for Particle {
  fn fmt(&self, f: &mut Formatter) -> Result {
    write!(
      f,
      "{{Particle mass: {} velocity: {} position: {}}}",
      self.mass,
      self.velocity,
      self.position
    )
  }
}
